compiler/lib/specialize.ml

```diff
@@ -117,3 +117,101 @@ let specialize_instrs ~function_arity p =
   { p with blocks; free_pc }
 
 let f = specialize_instrs
+
+(***)
+
+(* For switches, at this point, we know that this it is sufficient to
+   check the [pc]. *)
+let equal (pc, _) (pc', _) = pc = pc'
+
+let find_outlier_index arr =
+  let len = Array.length arr in
+  let rec find w i =
+    if i >= len
+    then `All_equals
+    else if equal arr.(i) w
+    then find w (i + 1)
+    else `Distinguished i
+  in
+  let a0 = arr.(0) in
+  match find a0 0 with
+  | `All_equals as res -> res
+  | `Distinguished i -> (
+      match find arr.(i) i with
+      | `All_equals ->
+          if i = 1
+          then `Distinguished 0
+          else if i = len - 1
+          then `Distinguished i
+          else `Splitted i
+      | `Distinguished j ->
+          if equal a0 arr.(j)
+          then
+            match find a0 j with
+            | `All_equals ->
+                if j = i + 1 then `Distinguished i else `Splitted_shifted (i, j)
+            | `Distinguished _ -> `Many_cases
+          else `Many_cases)
```

Comment:
I think you can skip the test `equal a0 arr.(j)` since `find a0 j` will return `Distinguished j` if the test is false:
```suggestion
            match find a0 j with
            | `All_equals ->
                if j = i + 1 then `Distinguished i else `Splitted_shifted (i, j)
            | `Distinguished _ -> `Many_cases
          )
```